node/core/pvf/src/worker_common.rs

```diff
@@ -0,0 +1,272 @@
+
+//! Common logic for implementation of worker processes.
+
+use crate::LOG_TARGET;
+use async_std::{
+	io,
+	os::unix::net::{UnixListener, UnixStream},
+	path::{PathBuf, Path},
+};
+use futures::{
+	AsyncRead, AsyncWrite, AsyncReadExt as _, AsyncWriteExt as _, FutureExt as _, never::Never,
+};
+use futures_timer::Delay;
+use rand::Rng;
+use std::{
+	fmt, mem,
+	pin::Pin,
+	task::{Context, Poll},
+	time::Duration,
+};
+use pin_project::pin_project;
+
+/// This is publicly exposed only for integration tests.
+#[doc(hidden)]
+pub async fn spawn_with_program_path(
+	debug_id: &'static str,
+	program_path: impl Into<PathBuf>,
+	extra_args: &'static [&'static str],
+	spawn_timeout: Duration,
+) -> Result<(IdleWorker, WorkerHandle), SpawnErr> {
+	let program_path = program_path.into();
+	with_transient_socket_path(debug_id, |socket_path| {
+		let socket_path = socket_path.to_owned();
+		async move {
+			let listener = UnixListener::bind(&socket_path)
+				.await
+				.map_err(|_| SpawnErr::Bind)?;
+
+			let handle = WorkerHandle::spawn(program_path, extra_args, socket_path)
+				.map_err(|_| SpawnErr::ProcessSpawn)?;
+
+			futures::select! {
+				accept_result = listener.accept().fuse() => {
+					let (stream, _) = accept_result.map_err(|_| SpawnErr::Accept)?;
+					Ok((IdleWorker { stream, pid: handle.id() }, handle))
+				}
+				_ = Delay::new(spawn_timeout).fuse() => {
+					Err(SpawnErr::AcceptTimeout)
+				}
+			}
+		}
+	})
+	.await
+}
+
+async fn with_transient_socket_path<R, F, Fut>(debug_id: &'static str, f: F) -> R
+where
+	F: FnOnce(&Path) -> Fut,
+	Fut: futures::Future<Output = R> + 'static,
+{
+	let socket_path = tmpfile(&format!("pvf-host-{}", debug_id));
+	let result = f(&socket_path).await;
+
+	// Best effort to remove the socket file. Under normal circumstances the socket will be removed
+	// by the worker. We make sure that it is removed here, just in case a failed rendezvous.
+	let _ = async_std::fs::remove_file(socket_path).await;
+
+	result
+}
+
+pub fn tmpfile(prefix: &str) -> PathBuf {
+	use rand::distributions::Alphanumeric;
+
+	const DESCRIMINATOR_LEN: usize = 10;
+
+	let mut buf = Vec::with_capacity(prefix.len() + DESCRIMINATOR_LEN);
+	buf.extend(prefix.as_bytes());
+	buf.extend(
+		rand::thread_rng()
+			.sample_iter(&Alphanumeric)
+			.take(DESCRIMINATOR_LEN),
+	);
+
+	let s = std::str::from_utf8(&buf).expect("the string is collected from a valid utf-8 sequence");
+
+	let mut temp_dir = PathBuf::from(std::env::temp_dir());
+	temp_dir.push(s);
+	temp_dir
+}
+
+pub fn worker_event_loop<F, Fut>(debug_id: &'static str, socket_path: &str, mut event_loop: F)
+where
+	F: FnMut(UnixStream) -> Fut,
+	Fut: futures::Future<Output = io::Result<Never>>,
+{
+	let err = async_std::task::block_on::<_, io::Result<()>>(async move {
+		let stream = UnixStream::connect(socket_path).await?;
+		let _ = async_std::fs::remove_file(socket_path).await;
+
+		event_loop(stream).await?;
+
+		Ok(())
+	})
+	.unwrap_err();
```

Comment:
Unwrap error? Hmm?

Reply:
oh, this is a remnant of previous version where it could only return with an error.

Reply:
ah, it's fine. the future's result is `Never`. I'll leave comment